Sources/Networking/Modifiers/Interceptors/Authorization/AuthorizationTokenInterceptor.swift

```diff
@@ -63,48 +67,24 @@ public extension AuthorizationTokenInterceptor {
 
 // MARK: Private methods
 private extension AuthorizationTokenInterceptor {
-    func performRefresh() async throws {
-        /// If some thread is already refreshing:
-        guard await !refreshingState.isRefreshing else {
-            /// Wait for signal to continue.
-            await refreshingState.wait()
-            await refreshingState.signal()
-            return
+    func refreshAuthorizationData() async throws {
+        /// In case the refresh is already in progress await it.
+        if let refreshTask = refreshTask {
+            return try await refreshTask.value
         }
-        
-        defer {
-            Task {
-                /// Unlock refreshing state and signal other threads that refreshing is done.
-                await refreshingState.setIsRefreshing(false)
-                await refreshingState.signal()
-            }
-        }
-        
-        /// Lock refreshing state to prevent other threads from trying to refresh as well.
-        await refreshingState.setIsRefreshing(true)
-        
-        /// Try refreshing authorization data.
-        try await authorizationManager.refreshAuthorizationData()
-    }
-}
 
-// MARK: Private actor
-private extension AuthorizationTokenInterceptor {
-    actor RefreshingState {
-        private let semaphore = AsyncSemaphore(value: 0)
-        
-        var isRefreshing = false
-        
-        func setIsRefreshing(_ value: Bool) {
-            isRefreshing = value
-        }
-        
-        func wait() async {
-            await semaphore.wait()
-        }
-        
-        func signal() {
-            semaphore.signal()
+        /// Otherwise create initiate new refresh task.
+        let newRefreshTask = Task { () throws -> Void in
+            /// Make sure to clear refreshTask property after refreshing finishes.
+            defer { refreshTask = nil }
```

Comment:
We retain strong self here

Reply:
Is it really necessary since it's an actor? 🤔 When I capture [weak self] in the Task and try to use the weak instance here, I get a compile error `Actor-isolated property 'refreshTask' can not be mutated from a non-isolated context`

Reply:
It is a problem, you retain `self` until the refresh task is finished which might potentially cause issues. You can fix the error like this: https://stackoverflow.com/questions/69259397/swift-actor-actor-isolated-property-scanning-can-not-be-mutated-from-a-non-i

Reply:
I see. Done ✅